qiskit_experiments/curve_analysis/base_curve_analysis.py

```diff
@@ -0,0 +1,550 @@
+
+
+class BaseCurveAnalysis(BaseAnalysis, ABC):
+    """Abstract superclass of curve analysis base classes.
+
+    Note that this class doesn't define :meth:`_run_analysis` method,
+    and no actual fitting protocol is implemented in this base class.
+    However, this class defines several common methods that can be reused.
+    A curve analysis subclass can construct proper fitting protocol
+    by combining following methods, i.e. subroutines.
+    See :ref:`curve_analysis_workflow` for how these subroutines are called.
+
+    .. rubric:: _generate_fit_guesses
+
+    An abstract method to create initial guesses for the fit parameters.
+    This should be implemented by subclass.
+    See :ref:`curve_analysis_init_guess` for details.
+
+    .. rubric:: _format_data
+
+    A method to format curve data. By default, this method takes the average of y values
+    over the same x values and then sort the entire data by x values.
+
+    .. rubric:: _evaluate_quality
+
+    A method to evaluate the quality of the fit based on the fit result.
+    This returns "good" when reduced chi-squared is less than 3.0.
+    Usually it returns string "good" or "bad" according to the evaluation.
+    This criterion can be updated by subclass.
+
+    .. rubric:: _run_data_processing
+
+    A method to perform data processing, i.e. create data arrays from
+    a list of experiment data payload.
+
+    .. rubric:: _run_curve_fit
+
+    A method to perform fitting with predefined fit models and formatted data.
+    This method internally calls :meth:`_generate_fit_guesses`.
+    Note that this is a core functionality of the :meth:`_run_analysis` method,
+    that creates fit result object from the processed curve data.
+
+    .. rubric:: _create_analysis_results
+
+    A method to create analysis results for important fit parameters
+    that might be defined by analysis options ``result_parameters``.
+    In addition, another entry for all fit parameters is created when
+    the analysis option ``return_fit_parameters`` is ``True``.
+
+    .. rubric:: _create_curve_data
+
+    A method to create analysis results for data points used for the fitting.
+    Entries are created when the analysis option ``return_data_points`` is ``True``.
+    If analysis consists of multiple series, analysis result is created for
+    each curve data in the series.
+
+    .. rubric:: _preparation
+
+    A method that should be called before other methods are called.
+    This method initializes analysis options against input experiment data.
+
+    """
+
+    def __init__(self):
+        """Initialize data fields that are privately accessed by methods."""
+        super().__init__()
+
+        #: List[int]: Index of physical qubits
+        self._physical_qubits = None
+
+    @property
+    @abstractmethod
+    def parameters(self) -> List[str]:
+        """Return parameters estimated by this analysis."""
+
+    @property
+    def drawer(self) -> BaseCurveDrawer:
```

Comment:
If its not for an end user, and just for internal coding in subclasses it should be `def _drawer`, though that is not much different than `self.options.curve_plotter` so I would argue isn't really needed at all

Reply:
This is also for the end users, i.e. I heard some people wanted to update drawer options more easily. With this one can update options with `exp.analysis.drawer.set_options` that looks good, however without this property one needs to call `exp.analysis.options.drawer.set_options` which looks bit awkward (option in the options).

Reply:
In that case it's a bit odd that the class is called CurveDrawer but the option is called `curve_plotter` I think naming should be consistent, both drawer or both plotter.